Editorial: define Math functions using algorithm steps

ryanjduffy opened the pull request:
Starting a few methods to get feedback.

Fixes #2119 

<!--
If you are changing the signature or behavior of an existing construct, please check if this affects downstream dependencies (searching for the construct's name is sufficient) and if needed file an issue:

* [Web IDL](https://heycam.github.io/webidl/) — [file an issue](https://github.com/heycam/webidl/issues/new)
* [HTML Standard](https://html.spec.whatwg.org/) — [file an issue](https://github.com/whatwg/html/issues/new)
-->

ljharb commented on spec.html:
```diff
@@ -27824,225 +27824,121 @@ <h1>Function Properties of the Math Object</h1>
-        </ul>
+        <emu-alg>
+            1. Let _n_ be ? ToNumber(_x_).
+            1. If Number.isNaN(_n_), return *NaN*.
```
this can't observably call into `Number.isNaN`; I think you may mean:
```suggestion
            1. If _n_ is *NaN*, return *NaN*.
```

ljharb commented on spec.html:
```diff
@@ -27824,225 +27824,121 @@ <h1>Function Properties of the Math Object</h1>
+            1. If Number.isNaN(_n_), return *NaN*.
+            1. If _n_ is *-0*, return *+0*.
+            1. If _n_ is *-&infin;*, return *+&infin;*.
+            1. If _n_ < 0, return ((_n_) * (-1)).
```
```suggestion
            1. If _n_ < 0, return -_n_.
```

ljharb commented on spec.html:
```diff
@@ -27824,225 +27824,121 @@ <h1>Function Properties of the Math Object</h1>
+            1. If _n_ is *-0*, return *+0*.
+            1. If _n_ is *-&infin;*, return *+&infin;*.
+            1. If _n_ < 0, return ((_n_) * (-1)).
+            1. Else return _n_.
```
```suggestion
            1. Return _n_.
```

ljharb commented on spec.html:
```diff
@@ -27824,225 +27824,121 @@ <h1>Function Properties of the Math Object</h1>
-        </ul>
+        <emu-alg>
+            1. Let _n_ be ? ToNumber(_x_).
+            1. If Number.isNaN(_n_) or _n_ &gt;1 or _n_ &lt; -1, return *NaN*.
```
```suggestion
            1. If _n_ is *NaN*, _n_ &gt; 1, or _n_ &lt; -1, return *NaN*.
```

ljharb replied:
(this still has two "or"s)

ryanjduffy commented:
Pushed a couple commits to apply @ljharb's feedback and generally:

* Separate conditions with commas (If _condition_, or _condition_, or _condition_, then ...`)
* Use `is *NaN*` over `Number.isNaN` and `is finite` over `Number.isFinite`

ljharb commented on spec.html:
```diff
@@ -27824,225 +27824,121 @@ <h1>Function Properties of the Math Object</h1>
-        </ul>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN*, or _n_ &gt;1, or _n_ &lt; -1, return *NaN*.
```
this also has two "or"s, as do a few others

ljharb replied:
Neither; i'm saying that in any comma-separated list, only the final item should have a conjunction.

gibson042 replied:
These could also be collapsed together, although that may not be true in every case.
```suggestion
          1. If _n_ is *NaN* or abs(_n_) &gt; 1, return *NaN*.
```

jmdyck replied:
> Neither; i'm saying that in any comma-separated list, only the final item should have a conjunction.

It sounds like you're saying to use (e.g.):
`If _n_ is 1, 2, or 3, ...`
rather than:
`If _n_ is 1, or 2, or 3, ...`
I'm fine with that rule (and it pretty much agrees with usage in the spec).

However, it doesn't apply to the current case, which has the structure
`If A, or B, or C, ...`
where each of A/B/C is a separate 'condition'. Currently, for such cases, the spec is somewhat split on whether to use any commas, but it's quite consistent that all the `or`s should appear.

(It's less clear when the conjunction is `and`:
`If A and B and C, ...`
is about as common as
`If A, B, and C, ...`
)

ljharb replied:
I wouldn't expect any differentiation between values or conditions; I'd expect what I'm used to in English sentences - that a comma-separated clause has only a final conjunction (and an Oxford comma, ofc).

jmdyck replied:
Well, then your expectation is at odds with current usage in the spec.

One downside of that style (in English and in the spec) is that you have to read most of the condition before you know whether the elided conjunction is "and" or "or". Which is not a big problem if the condition is short, but the spec has some long ones.

ryanjduffy replied:
In this PR, I've used the following format but I can adjust as needed:

```
1. If condition1, then ...
1. If condition1 or condition2, then ...
1. If condition1, condition2, or condition3, then ...
```

michaelficarra commented on spec.html:
```diff
@@ -28586,29 +28371,14 @@ <h1>Math.tanh ( _x_ )</h1>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN*, _n_ is *-0*, _n_ is *+0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
+          1. If _n_ &gt; 0 and _n_ &lt; 1, return *+0*.
+          1. If _n_ &lt; 0 and _n_ &lt;= -1, return *-0*.
```
```suggestion
          1. If _n_ &lt; 0 and _n_ &gt; -1, return *-0*.
```

michaelficarra commented on spec.html:
```diff
@@ -28586,29 +28371,14 @@ <h1>Math.tanh ( _x_ )</h1>
+          1. If _n_ &gt; 0 and _n_ &lt; 1, return *+0*.
+          1. If _n_ &lt; 0 and _n_ &lt;= -1, return *-0*.
+          1. Let _r_ be _n_ modulo 1.
+          1. Return _n_ - _r_.
```
This doesn't work for negative `x`, right? I think you'll have to say something like

```suggestion
          1. If n &gt; 0, return _n_ - _r_; otherwise, return _n_ + (1 - _r_).
```

ryanjduffy replied:
Yep!

michaelficarra commented on spec.html:
```diff
@@ -28586,29 +28371,14 @@ <h1>Math.tanh ( _x_ )</h1>
+          1. If _n_ is *NaN*, _n_ is *-0*, _n_ is *+0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
+          1. If _n_ &gt; 0 and _n_ &lt; 1, return *+0*.
+          1. If _n_ &lt; 0 and _n_ &lt;= -1, return *-0*.
+          1. Let _r_ be _n_ modulo 1.
+          1. Return _n_ - _r_.
```
Something like this might be even better (due to float maths weirdness):

```suggestion
          1. Return the integral Number nearest _n_ in the direction of *+0*.
```

@bakkot Thoughts?

ryanjduffy replied:
I went with this since it aligns with the changes to `Math.round()`.

michaelficarra commented on spec.html:
```diff
@@ -28420,29 +28274,19 @@ <h1>Math.random ( )</h1>
+          1. If _n_ is *NaN*, _n_ is *-0*, _n_ is *+0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
+          1. If _n_ &gt; 0 and _n_ &lt; 0.5, return *+0*.
+          1. If _n_ &lt; 0 and _n_ &lt;= -0.5, return *-0*.
+          1. Let _r_ be _n_ modulo 1.
+          1. Let _v_ be _n_ - _r_.
+          1. If _r_ is 0, return _v_.
+          1. If _r_ &gt;= -0.5 and _r_ &lt; 0, return _v_.
+          1. If _r_ &lt; 0.5 and _r_ &gt; 0, return _v_.
+          1. If _r_ &lt; -0.5, return _v_ - 1.
+          1. Return _v_ + 1.
```
I think we can say this more simply and equally clearly with prose:

```suggestion
          1. If _n_ is an integral Number, return _n_.
          1. Return the integral Number closest to _n_, preferring the Number closer to *+&infin;* in the case of a tie.
```

ryanjduffy replied:
That's fine. It was unclear to me to how close these should be to implementation algorithms so I probably went too far on this one.

michaelficarra commented on spec.html:
```diff
@@ -28276,130 +28172,88 @@ <h1>Math.imul ( _x_, _y_ )</h1>
-        </ul>
+        <emu-alg>
+          1. Let _values_ be a List whose elements are, in left to right order, the arguments that were passed to this function invocation.
+          1. Let _len_ be the number of elements in _values_.
```
This isn't used.

```suggestion
```

michaelficarra commented on spec.html:
```diff
@@ -28276,130 +28172,88 @@ <h1>Math.imul ( _x_, _y_ )</h1>
-          </li>
-        </ul>
+        <emu-alg>
+          1. Let _values_ be a List whose elements are, in left to right order, the arguments that were passed to this function invocation.
```
I don't know if you need this line given that `values` is declared in the parameter list. @tc39/ecma262-editors?

ljharb replied:
agreed; it's redundant with `...values` in the signature.

ryanjduffy replied:
I borrowed this from `Array.prototype.push` and `Array.prototype.splice` which I only point out so they can also be updated separately if that's the desire. I'll drop this here.

michaelficarra replied:
We'll discuss this in https://github.com/tc39/ecma262/pull/2139.

michaelficarra commented on spec.html:
```diff
@@ -28276,130 +28172,88 @@ <h1>Math.imul ( _x_, _y_ )</h1>
+          1. Let _len_ be the number of elements in _values_.
+          1. Let _v_ be *-&infin;*.
+          1. Repeat, while _values_ is not empty,
+            1. Remove the first element from _values_ and let _value_ be the value of the element.
```
I would just make this two steps.

```suggestion
            1. Let _value_ be the first element of _values_.
            1. Remove the first element from _values_.
```

ryanjduffy replied:
This format is also used in `Array` methods (e.g. `Array.prototype.concat`) which is from where I borrowed it.

michaelficarra replied:
Thanks, I've sent PR https://github.com/tc39/ecma262/pull/2137 to address those cases.

michaelficarra commented on spec.html:
```diff
@@ -28276,130 +28172,88 @@ <h1>Math.imul ( _x_, _y_ )</h1>
+        <emu-alg>
+          1. Let _values_ be a List whose elements are, in left to right order, the arguments that were passed to this function invocation.
+          1. Let _len_ be the number of elements in _values_.
+          1. Let _v_ be *-&infin;*.
```
I'd probably use a better alias name here. Something like `lowest`.

michaelficarra replied:
If you end up not changing the signature, you can do something like this:

```suggestion
          1. Let _v_ be *-&infin;*.
          1. If _value1_ &gt; _v_, set _v_ to _value1_.
          1. If _value2_ &gt; _v_, set _v_ to _value2_.
```

michaelficarra commented on spec.html:
```diff
@@ -28276,130 +28172,88 @@ <h1>Math.imul ( _x_, _y_ )</h1>
 
       <emu-clause id="sec-math.max">
-        <h1>Math.max ( _value1_, _value2_, ..._values_ )</h1>
+        <h1>Math.max ( ..._values_ )</h1>
```
It may be surprising, but this change is actually normative. `Math.max.length` is defined to be `2` because of the number of mandatory parameters listed here. We can resolve this by explicitly stating that the length of this function is `2` in the description. Or you could change the signature back. Either way works for me.

ljharb replied:
I have a strong preference to change the signature back, and to minimize the places in the spec where the length is explicitly overriding the signature.

ryanjduffy replied:
Agreed!

michaelficarra commented on spec.html:
```diff
@@ -28276,130 +28172,88 @@ <h1>Math.imul ( _x_, _y_ )</h1>
-        </ul>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN* or _n_ &lt; 0, return *NaN*.
+          1. If _n_ is *+0* or _n_ is *-0*, return *-&infin;*.
+          1. If _n_ is 1, return *+0*.
+          1. If _n_ is *+&infin;*, return *+&infin;*.
```
I slightly prefer grouping together the pass-throughs like the first line here:

```suggestion
          1. If _n_ is *NaN* or _n_ is *+&infin;*, return _n_.
          1. If _n_ &lt; 0, return *NaN*.
          1. If _n_ is *+0* or _n_ is *-0*, return *-&infin;*.
          1. If _n_ is 1, return *+0*.
```

michaelficarra commented on spec.html:
```diff
@@ -27824,225 +27824,121 @@ <h1>Function Properties of the Math Object</h1>
-        </ul>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN*, return *NaN*.
+          1. If _n_ is *+0*, _n_ is *-0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
```
```suggestion
          1. If _n_ is *NaN*, _n_ is *+0*, _n_ is *-0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
```

ryanjduffy commented:
@michaelficarra Thanks for the feedback. Hoping to get back to this later tonight.

michaelficarra commented:
@ryanjduffy Is this ready for another review?

ryanjduffy commented:
Finished converting the methods but haven't reviewed the code yet. Reviews are still welcome but I won't move this out of draft until I've had a chance to review in detail.

cc: @ljharb, @michaelficarra

ryanjduffy commented:
@michaelficarra - just saw your note. Yes, I can do that.

michaelficarra commented:
@ryanjduffy We should remove these paragraphs at the top of 20.3.2 as part of this PR:

![image](https://user-images.githubusercontent.com/218840/90460084-fad32500-e0b7-11ea-9089-11ac2e94f3f6.png)

michaelficarra commented on spec.html:
```diff
@@ -28130,273 +28130,144 @@ <h1>Function Properties of the Math Object</h1>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN*, _n_ is *+0*, _n_ is *-0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
+          1. If _n_ &lt; 0 and _n_ &gt; -1, return *-0*.
+          1. If _n_ is an integer, return _n_.
```
We try to use the term "integral Number" for Number values that represent a mathematical integer.

```suggestion
          1. If _n_ is an integral Number, return _n_.
```

michaelficarra commented on spec.html:
```diff
@@ -28130,273 +28130,144 @@ <h1>Function Properties of the Math Object</h1>
+          1. If _n_ is *NaN*, _n_ is *+0*, _n_ is *-0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
+          1. If _n_ &lt; 0 and _n_ &gt; -1, return *-0*.
+          1. If _n_ is an integer, return _n_.
+          1. Return the smallest (closest to *-&infin;*) Number value that is not less than _n_ and is an integer.
```
```suggestion
          1. Return the smallest (closest to *-&infin;*) integral Number value that is not less than _n_.
```

ryanjduffy commented:
Updated "integer" to "integral Number" in `floor()` and `ceil()` and removed those two paragraphs.

michaelficarra commented on spec.html:
```diff
@@ -28546,23 +28359,27 @@ <h1>Math.fround ( _x_ )</h1>
+            1. If _v_ is not *+0* or _v_ is not *-0*, set _onlyZero_ to be *false*.
+          1. If _hasNaN_ is *true*, return *NaN*.
+          1. If _onlyZero_ is *true*, return *+0*.
+          1. Return an implementation-approximated value representing the result of the square root of the sum of squares of its arguments.
```
I don't think this step is sufficient. I think you should construct a List of all the `ToNumber` results like

```
1. Let _numbers_ be a List containing _v1_ and _v2_.
...
1. Repeat, while _values_ is not empty,
  ...
  1. Append _v_ to _numbers_.
```

and then say here something like

```
1. Return an implementation-approximated value representing the square root of the sum of squares of the elements of _numbers_.
```

Does that make sense?

ryanjduffy replied:
I took a slightly different approach that suggested but let me know if it doesn't work or follow convention.

```
        <emu-alg>
          1. Let _numbers_ be a List containing _value1_ and _value2_, and the elements of _values_ in order.
          1. Let _onlyZero_ be *true*.
          1. Repeat, while _numbers_ is not empty,
            1. Let _number_ be the first element of _numbers_.
            1. Remove the first element from _numbers_.
            1. Let _n_ be ? ToNumber(_number_).
            1. If _n_ is *NaN*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
            1. If _n_ is not *+0* or _n_ is not *-0*, set _onlyZero_ to be *false*.
          1. If _onlyZero_ is *true*, return *+0*.
          1. Return an implementation-approximated value representing the result of the square root of the sum of squares of the elements of _numbers_.
        </emu-alg>
```

I used the similar `List` construction sentence proposed below of all the values. Since NaN and +/- infinity all cause a bail out, I thought it was easier to grok by iterating over the original list and returning the exception cases.

michaelficarra commented on spec.html:
```diff
@@ -28546,23 +28359,27 @@ <h1>Math.fround ( _x_ )</h1>
-          </li>
-        </ul>
+        <emu-alg>
+          1. Let _v1_ be ? ToNumber(_value1_).
+          1. Let _v2_ be ? ToNumber(_value2_).
+          1. If _v1_ is *NaN* or _v2_ is *NaN*, return *NaN*.
+          1. If _v1_ is *+&infin;* or _v2_ is *+&infin;*, return *+&infin;*.
+          1. If _v1_ is *-&infin;* or _v2_ is *-&infin;*, return *-&infin;*.
+          1. Let _hasNaN_ be *false*.
+          1. Let _onlyZero_ be *true*.
+          1. If _v1_ is not 0 and _v2_ is not 0, set _onlyZero_ to *false*.
```
I think we can skip all of this duplication by constructing a list like

```
1. Let _args_ be a List containing _value1_, _value2_, and the elements of _values_ in order.
```

and then just looping over `_args_` below instead of `_values_`.

michaelficarra commented on spec.html:
```diff
@@ -28582,130 +28399,89 @@ <h1>Math.imul ( _x_, _y_ )</h1>
-        </ul>
+        <emu-alg>
+          1. Let _highest_ be *-&infin;*.
+          1. If _value1_ &gt; _highest_, set _highest_ to _value1_.
```
We're forgetting to call `ToNumber` on `_value1_` and `_value2_` here. I'm thinking it might be easiest to prepend them to `_values_` and re-use the loop steps below, as I suggested for `Math.hypot`.

michaelficarra commented on spec.html:
```diff
@@ -28582,130 +28399,89 @@ <h1>Math.imul ( _x_, _y_ )</h1>
-        </ul>
+        <emu-alg>
+          1. Let _lowest_ be *+&infin;*.
+          1. If _value1_ &lt; _lowest_, set _lowest_ to _value1_.
```
Same comment as `Math.max`.

ryanjduffy commented:
I think I've addressed all the feedback so I've moved this out of draft. Thanks for the help getting up to speed with the conventions, @michaelficarra  and @ljharb!

michaelficarra commented on spec.html:
```diff
@@ -28360,25 +28360,19 @@ <h1>Math.fround ( _x_ )</h1>
           1. If _hasNaN_ is *true*, return *NaN*.
           1. If _onlyZero_ is *true*, return *+0*.
-          1. Return an implementation-approximated value representing the result of the square root of the sum of squares of its arguments.
+          1. Return an implementation-approximated value representing the result of the square root of the sum of squares of the elements of _numbers_.
```
Nit:

```suggestion
          1. Return an implementation-approximated value representing the square root of the sum of squares of the elements of _numbers_.
```

Same in the summary.

michaelficarra commented on spec.html:
```diff
@@ -28346,7 +28361,7 @@ <h1>Math.floor ( _x_ )</h1>
       <emu-clause id="sec-math.fround">
         <h1>Math.fround ( _x_ )</h1>
-        <p>When `Math.fround` is called with argument _x_, the following steps are taken:</p>
+        <p>When the `Math.fround` method is called with argument _x_, the following steps are taken:</p>
```
Every other unary function's summary now says "one argument", so let's keep it consistent.

```suggestion
        <p>When the `Math.fround` method is called with one argument _x_, the following steps are taken:</p>
```

ljharb commented on spec.html:
```diff
@@ -28132,278 +28130,159 @@ <h1>Function Properties of the Math Object</h1>
+            1. Let _n_ be ? ToNumber(_x_).
+            1. If _n_ is *NaN*, _n_ &gt; 1, or _n_ &lt; -1, return *NaN*.
+            1. If _n_ is 1, return *+0*.
+            1. Return an implementation-approximated value representing the result of the arc cosine of _n_.
```
```suggestion
            1. Return an implementation-approximated value representing the result of the inverse cosine of _n_.
```

ljharb commented on spec.html:
```diff
@@ -28132,278 +28130,159 @@ <h1>Function Properties of the Math Object</h1>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN*, _n_ is *+0*, or _n_ is *-0*, return _n_.
+          1. If _n_ &gt; 1 or _n_ &lt; -1, return *NaN*.
+          1. Return an implementation-approximated value representing the result of the arc sine of _n_.
```
```suggestion
          1. Return an implementation-approximated value representing the result of the inverse sine of _n_.
```

ljharb commented on spec.html:
```diff
@@ -28132,278 +28130,159 @@ <h1>Function Properties of the Math Object</h1>
+          1. If _n_ is *NaN*, _n_ is *+0*, or _n_ is *-0*, return _n_.
+          1. If _n_ is *+&infin;*, return an implementation-approximated value representing +&pi; / 2.
+          1. If _n_ is *-&infin;*, return an implementation-approximated value representing -&pi; / 2.
+          1. Return an implementation-approximated value representing the result of the arc tangent of _n_.
```
```suggestion
          1. Return an implementation-approximated value representing the result of the inverse tangent of _n_.
```

ljharb commented on spec.html:
```diff
@@ -28132,278 +28130,159 @@ <h1>Function Properties of the Math Object</h1>
+          1. If _ny_ is *-&infin;*, then
+            1. If _nx_ is *+&infin;*, return an implementation-approximated value representing -&pi; / 4.
+            1. If _nx_ is *-&infin;*, return an implementation-approximated value representing -3&pi; / 4.
+          1. Return an implementation-approximated value representing the result of the arc tangent of the quotient _ny_ / _nx_.
```
```suggestion
          1. Return an implementation-approximated value representing the result of the inverse tangent of the quotient _ny_ / _nx_.
```

ljharb replied:
I think that this PR needs to be overall consistent with either using "arc" or "inverse"; currently it uses both.

ljharb commented on spec.html:
```diff
@@ -28464,78 +28323,47 @@ <h1>Math.cosh ( _x_ )</h1>
       <emu-clause id="sec-math.fround">
         <h1>Math.fround ( _x_ )</h1>
-        <p>When `Math.fround` is called with argument _x_, the following steps are taken:</p>
+        <p>When the `Math.fround` method is called with one argument _x_, the following steps are taken:</p>
```
```suggestion
        <p>When the `Math.fround` method is called with one argument _x_, the following steps are taken:</p>
```
or perhaps, remove "one" from all these preambles? "is called with one" appears 36 times in the spec, and "is called with arg" appears 113 times, so it seems like it'd be more consistent to go with the more common form and make them all "is called with argument".

(Additionally, by specifying the number of arguments there's a mild implication that if it's called with zero or 2+ args, different steps are taken, which isn't the case)

ljharb replied:
Yes, I would expect and prefer that all these explicit mentions of a count followed by a short list where the reader can easily count for themselves should remove the count :-)

ljharb commented on spec.html:
```diff
@@ -28728,29 +28517,12 @@ <h1>Math.random ( )</h1>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is an integral Number, return _n_.
+          1. Return the integral Number closest to _n_, preferring the Number closer to *+&infin;* in the case of a tie.
```
it seems important not to lose the explicit mention of `0.5` here as part of the rounding method.

michaelficarra replied:
Do you have a concrete suggestion? I personally think the current text is sufficient.

ljharb replied:
```suggestion
          1. If _x_ &gt; 0 and _x_ &lt; 0.5, return *+0*.
          1. If _x_ &lt; 0 and _x_ &ge; -0.5, return *-0*.
          1. Return the integral Number closest to _n_, preferring the Number closer to *+&infin;* in the case of a tie.
```

michaelficarra replied:
Either is fine by me.

syg commented on spec.html:
```diff
@@ -28132,278 +28130,159 @@ <h1>Function Properties of the Math Object</h1>
 
       <emu-clause id="sec-math.acos">
         <h1>Math.acos ( _x_ )</h1>
         <p>Returns an implementation-approximated value representing the result of the arc cosine of _x_. The result is expressed in radians and ranges from *+0* to +&pi;.</p>
```
If we're unifying on "inverse" instead of "arc" here and below, this paragraph should change too.

syg commented on spec.html:
```diff
@@ -28132,278 +28130,159 @@ <h1>Function Properties of the Math Object</h1>
 
       <emu-clause id="sec-math.asin">
         <h1>Math.asin ( _x_ )</h1>
         <p>Returns an implementation-approximated value representing the result of the arc sine of _x_. The result is expressed in radians and ranges from -&pi; / 2 to +&pi; / 2.</p>
```
s/arc/inverse

syg commented on spec.html:
```diff
@@ -28132,278 +28130,159 @@ <h1>Function Properties of the Math Object</h1>
 
       <emu-clause id="sec-math.atan">
         <h1>Math.atan ( _x_ )</h1>
         <p>Returns an implementation-approximated value representing the result of the arc tangent of _x_. The result is expressed in radians and ranges from -&pi; / 2 to +&pi; / 2.</p>
```
s/arc/inverse

ryanjduffy commented:
@syg, @michaelficarra  - I simplified the summaries (in 5a29038) in cases where it seemed redundant to the steps. I think it reads better but would appreciate your thoughts.

michaelficarra commented on spec.html:
```diff
@@ -28168,7 +28168,7 @@ <h1>Math.acosh ( _x_ )</h1>
       <emu-clause id="sec-math.asin">
         <h1>Math.asin ( _x_ )</h1>
-        <p>Returns an implementation-approximated value representing the result of the arc sine of _x_. The result is expressed in radians and ranges from -&pi; / 2 to +&pi; / 2.</p>
+        <p>Returns the inverse sine of _x_. The argument is expressed in radians and ranges from -&pi; / 2 to +&pi; / 2.</p>
```
You changed a lot of these to "the argument", but I think they're supposed to be talking about the result.

ryanjduffy replied:
My mistake! Got a little carried away in the standardization. :/

ryanjduffy replied:
Fixed this up!

michaelficarra commented on spec.html:
```diff
@@ -28375,7 +28375,7 @@ <h1>Math.fround ( _x_ )</h1>
       <emu-clause id="sec-math.hypot">
         <h1>Math.hypot ( _value1_, _value2_, ..._values_ )</h1>
-        <p>`Math.hypot` returns an implementation-approximated value representing the square root of the sum of squares of its arguments.</p>
+        <p>`Math.hypot` returns the square root of the sum of squares of its arguments.</p>
```
Nit: this summary stands out in that it mentions the method name again at the beginning. The others just start with "Returns ...".

ryanjduffy replied:
Fixed!

michaelficarra commented:
@ryanjduffy FYI, I pushed a commit to align with #2152.

ljharb commented on spec.html:
```diff
@@ -28538,135 +28360,96 @@ <h1>Math.imul ( _x_, _y_ )</h1>
-        </ul>
+        <p>When the `Math.max` method is called with at least two arguments _value1_ and _value2_ and any number of additional arguments which form the rest parameter ..._values_, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _numbers_ be a List containing _value1_, _value2_, and the elements of _values_ in order.
```
```suggestion
          1. Let _numbers_ be a List containing _value1_, _value2_, and the elements of _values_ in List order.
```

ljharb commented on spec.html:
```diff
@@ -28538,135 +28360,96 @@ <h1>Math.imul ( _x_, _y_ )</h1>
-        </ul>
+        <p>When the `Math.min` method is called with at least two arguments _value1_ and _value2_ and any number of additional arguments which form the rest parameter ..._values_, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _numbers_ be a List containing _value1_, _value2_, and the elements of _values_ in order.
```
```suggestion
          1. Let _numbers_ be a List containing _value1_, _value2_, and the elements of _values_ in List order.
```

ryanjduffy commented:
@ljharb - Shall I make some updates based on your feedback or do you plan to take care of them since you self-assigned?

ljharb commented:
@ryanjduffy none of the suggestions are blockers; feel free to make them yourself :-) i self-assigned since i plan to merge this once enough editors have stamped.

ryanjduffy commented:
Updated!

bakkot commented on spec.html:
```diff
@@ -28371,126 +28250,75 @@ <h1>Math.clz32 ( _x_ )</h1>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN* or _n_ is *+&infin;*, return _n_.
+          1. If _n_ is *-0* or _n_ is *+0*, return 1.
```
```suggestion
          1. If _n_ is *+0* or _n_ is *-0*, return 1.
```

bakkot commented on spec.html:
```diff
@@ -28371,126 +28250,75 @@ <h1>Math.clz32 ( _x_ )</h1>
+        <p>When the `Math.expm1` method is called with argument _x_, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN*, _n_ is *-0*, _n_ is *+0*, or _n_ is *+&infin;*, return _n_.
```
```suggestion
          1. If _n_ is *NaN*, _n_ is *+0*, _n_ is *-0*, or _n_ is *+&infin;*, return _n_.
```

bakkot commented on spec.html:
```diff
@@ -28502,24 +28330,18 @@ <h1>Math.fround ( _x_ )</h1>
+          1. For each element _number_ of _numbers_, do
+            1. Let _n_ be ? ToNumber(_number_).
+            1. If _n_ is *NaN*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
+            1. If _n_ is not *+0* or _n_ is not *-0*, set _onlyZero_ to be *false*.
```
```suggestion
            1. If _n_ is neither *+0* nor *-0*, set _onlyZero_ to *false*.
```

michaelficarra replied:
Nice catch!

bakkot commented on spec.html:
```diff
@@ -28538,135 +28360,96 @@ <h1>Math.imul ( _x_, _y_ )</h1>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN*, _n_ is *+0*, _n_ is *-0*, or _n_ is *+&infin;*, return _n_.
+          1. If _n_ &lt; -1, return *NaN*.
+          1. If _n_ is -1, return *-&infin;*.
```
Better to go in descending order:
```suggestion
          1. If _n_ is -1, return *-&infin;*.
          1. If _n_ &lt; -1, return *NaN*.
```

bakkot commented on spec.html:
```diff
@@ -28538,135 +28360,96 @@ <h1>Math.imul ( _x_, _y_ )</h1>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN* or _n_ is *+&infin;*, return _n_.
+          1. If _n_ &lt; 0, return *NaN*.
+          1. If _n_ is *+0* or _n_ is *-0*, return *-&infin;*.
+          1. If _n_ is 1, return *+0*.
```
Better to go in descending order:
```suggestion
          1. If _n_ is 1, return *+0*.
          1. If _n_ is *+0* or _n_ is *-0*, return *-&infin;*.
          1. If _n_ &lt; 0, return *NaN*.
```

ryanjduffy replied:
I had organized the steps to list all "identity" cases, followed by exceptional cases (e.g. inputs which return NaN or that have specific return values) and then the "regular" cases. 

How do you want to handle cases like `Math.atanh`?

```
          1. Let _n_ be ? ToNumber(_x_).
          1. If _n_ is *NaN*, _n_ is *+0*, or _n_ is *-0*, return _n_.
          1. If _n_ &gt; 1 or _n_ &lt; -1, return *NaN*.
          1. If _n_ is +1, return *+&infin;*.
          1. If _n_ is -1, return *-&infin;*.
          1. Return an implementation-approximated value representing the result of the inverse hyperbolic tangent of _n_.
```

To strictly follow your guidance, I could split the third step so that the order is > 1, = 1, = -1, < -1. Alternatively this could remain as is and we treat multiple conditions as special cases that precede the descending order. Probably other options too. What do you think, @bakkot?

michaelficarra replied:
@bakkot See https://github.com/tc39/ecma262/pull/2122#discussion_r465876073.

bakkot replied:
> How do you want to handle cases like `Math.atanh`?

In cases where there's no natural order because some conditions are combined this preference wouldn't apply. I only intended this comment to apply to the `log` methods, and only for these three cases (1, 0, <0).

@michaelficarra I saw that comment, but I don't understand the relevance here.

ryanjduffy replied:
@bakkot - Thanks! I've made a few related changes to prefer descending order (e.g. [this one](https://github.com/tc39/ecma262/pull/2122/commits/e7c17df9c70b0575e0d483fc09b67fbeb50a452d#diff-3540caefa502006d8a33cb1385720803R28309)) but the `log` entries were indeed the primary ones.

bakkot commented on spec.html:
```diff
@@ -28538,135 +28360,96 @@ <h1>Math.imul ( _x_, _y_ )</h1>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN* or _n_ is *+&infin;*, return _n_.
+          1. If _n_ &lt; 0, return *NaN*.
+          1. If _n_ is *+0* or _n_ is *-0*, return *-&infin;*.
+          1. If _n_ is 1, return *+0*.
```
```suggestion
          1. If _n_ is 1, return *+0*.
          1. If _n_ is *+0* or _n_ is *-0*, return *-&infin;*.
          1. If _n_ &lt; 0, return *NaN*.
```

bakkot commented on spec.html:
```diff
@@ -28538,135 +28360,96 @@ <h1>Math.imul ( _x_, _y_ )</h1>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN* or _n_ is *+&infin;*, return _n_.
+          1. If _n_ &lt; 0, return *NaN*.
+          1. If _n_ is *+0* or _n_ is *-0*, return *-&infin;*.
+          1. If _n_ is 1, return *+0*.
```
```suggestion
          1. If _n_ is 1, return *+0*.
          1. If _n_ is *+0* or _n_ is *-0*, return *-&infin;*.
          1. If _n_ &lt; 0, return *NaN*.
```

bakkot commented on spec.html:
```diff
@@ -28538,135 +28360,96 @@ <h1>Math.imul ( _x_, _y_ )</h1>
+          1. Let _highest_ be *-&infin;*.
+          1. For each element _number_ of _numbers_, do
+            1. Let _n_ be ? ToNumber(_number_).
+            1. If _n_ &gt; _highest_, set _highest_ to _n_.
```
We need to explicitly treat `NaN`, `+0`, and `-0` here, I think; it does not make sense to ask if `NaN > 1`, and we have to ensure that `+0` is larger than `-0`.

ryanjduffy replied:
How does this look? Would it be better to break the second link into two separate conditions or somehow group the two conditions joined with "and" for better clarity?

```
            1. If _n_ is *NaN*, return *NaN*.
            1. If _n_ &gt; _highest_ or _n_ is +0 and _highest_ is *-0*, set _highest_ to _n_.
```

ryanjduffy replied:
Maybe this is a bit better

```
            1. If _n_ is *NaN*, return *NaN*.
            1. If _n_ is +0 and _highest_ is *-0* or _n_ &gt; _highest_, set _highest_ to _n_.
```

michaelficarra replied:
@ryanjduffy Please don't combine things like that. The precedence of and/or in English is ambiguous.

ryanjduffy replied:
:+1: That was my instinct as well. I'll separate them out.

ryanjduffy replied:
I've split them into separate lines. Seems to read better but open to suggestions as well.

bakkot commented on spec.html:
```diff
@@ -28538,135 +28360,96 @@ <h1>Math.imul ( _x_, _y_ )</h1>
+          1. Let _lowest_ be *+&infin;*.
+          1. For each element _number_ of _numbers_, do
+            1. Let _n_ be ? ToNumber(_number_).
+            1. If _n_ &lt; _lowest_, set _lowest_ to _n_.
```
As above, we need to explicitly treat `NaN`, `+0`, and `-0` here.

bakkot commented on spec.html:
```diff
@@ -28717,130 +28485,76 @@ <h1>Math.round ( _x_ )</h1>
+        <p>When the `Math.sign` method is called with argument _x_, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN*, _n_ is *-0*, or _n_ is *+0*, return _n_.
```
```suggestion
          1. If _n_ is *NaN*, _n_ is *+0*, or _n_ is *-0*, return _n_.
```

bakkot commented on spec.html:
```diff
@@ -28717,130 +28485,76 @@ <h1>Math.round ( _x_ )</h1>
+        <p>When the `Math.sin` method is called with argument _x_, the following steps are taken:</p>
+        <emu-alg>
+            1. Let _n_ be ? ToNumber(_x_).
+            1. If _n_ is *NaN*, _n_ is *-0*, or _n_ is *+0*, return _n_.
```
```suggestion
            1. If _n_ is *NaN*, _n_ is *+0*, or _n_ is *-0*, return _n_.
```

bakkot commented on spec.html:
```diff
@@ -28717,130 +28485,76 @@ <h1>Math.round ( _x_ )</h1>
+        <emu-alg>
+            1. Let _n_ be ? ToNumber(_x_).
+            1. If _n_ is *NaN*, _n_ is *-0*, or _n_ is *+0*, return _n_.
+            1. If _n_ is *-&infin;* or _n_ is *+&infin;*, return *NaN*.
```
```suggestion
            1. If _n_ is *+&infin;* or _n_ is *-&infin;*, return *NaN*.
```

bakkot commented on spec.html:
```diff
@@ -28717,130 +28485,76 @@ <h1>Math.round ( _x_ )</h1>
+        <p>When the `Math.sinh` method is called with argument _x_, the following steps are taken:</p>
+        <emu-alg>
+            1. Let _n_ be ? ToNumber(_x_).
+            1. If _n_ is *NaN*, _n_ is *-0*, _n_ is *+0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
```
```suggestion
            1. If _n_ is *NaN*, _n_ is *+0*, _n_ is *-0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
```

bakkot commented on spec.html:
```diff
@@ -28717,130 +28485,76 @@ <h1>Math.round ( _x_ )</h1>
+        <p>When the `Math.sqrt` method is called with argument _x_, the following steps are taken:</p>
+        <emu-alg>
+            1. Let _n_ be ? ToNumber(_x_).
+            1. If _n_ is *NaN*, _n_ is *-0*, _n_ is *+0*, or _n_ is *+&infin;*, return _n_.
```
```suggestion
            1. If _n_ is *NaN*, _n_ is *+0*, _n_ is *-0*, or _n_ is *+&infin;*, return _n_.
```

bakkot commented on spec.html:
```diff
@@ -28717,130 +28485,76 @@ <h1>Math.round ( _x_ )</h1>
+        <p>When the `Math.tan` method is called with argument _x_, the following steps are taken:</p>
+        <emu-alg>
+            1. Let _n_ be ? ToNumber(_x_).
+            1. If _n_ is *NaN*, _n_ is *-0*, or _n_ is *+0*, return _n_.
```
```suggestion
            1. If _n_ is *NaN*, _n_ is *+0*, or _n_ is *-0*, return _n_.
```

bakkot commented on spec.html:
```diff
@@ -28717,130 +28485,76 @@ <h1>Math.round ( _x_ )</h1>
+        <emu-alg>
+            1. Let _n_ be ? ToNumber(_x_).
+            1. If _n_ is *NaN*, _n_ is *-0*, or _n_ is *+0*, return _n_.
+            1. If _n_ is *-&infin;*, or _n_ is *+&infin;*, return *NaN*.
```
```suggestion
            1. If _n_ is *+&infin;* or _n_ is *-&infin;*, return *NaN*.
```

(note also the stray comma)

bakkot commented on spec.html:
```diff
@@ -28717,130 +28485,76 @@ <h1>Math.round ( _x_ )</h1>
+        <p>When the `Math.tanh` method is called with argument _x_, the following steps are taken:</p>
+        <emu-alg>
+            1. Let _n_ be ? ToNumber(_x_).
+            1. If _n_ is *NaN*, _n_ is *-0*, or _n_ is *+0*, return _n_.
```
```suggestion
            1. If _n_ is *NaN*, _n_ is *+0*, or _n_ is *-0*, return _n_.
```

bakkot commented on spec.html:
```diff
@@ -28849,29 +28563,14 @@ <h1>Math.tanh ( _x_ )</h1>
+        <p>When the `Math.trunc` method is called with argument _x_, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _n_ be ? ToNumber(_x_).
+          1. If _n_ is *NaN*, _n_ is *-0*, _n_ is *+0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
```
```suggestion
          1. If _n_ is *NaN*, _n_ is *+0*, _n_ is *-0*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
```

bakkot reviewed (commented):
Looks pretty good. I had a few small comments.

ryanjduffy commented on spec.html:
```diff
@@ -28087,278 +28085,159 @@ <h1>Function Properties of the Math Object</h1>
+        <p>Returns the inverse tangent of the quotient <emu-eqn>_y_ / _x_</emu-eqn> of the arguments _y_ and _x_, where the signs of _y_ and _x_ are used to determine the quadrant of the result. Note that it is intentional and traditional for the two-argument inverse tangent function that the argument named _y_ be first and the argument named _x_ be second. The result is expressed in radians and  ...
+        <p>When the `Math.atan2` method is called with arguments _x_ and _y_, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _ny_ be ? ToNumber(_y_).
```
Thanks! Agree this is more readable.

ryanjduffy commented:
Thanks for the feedback, @bakkot. Had a couple questions I shared above but I pushed a batch changes to keep things moving.

jmdyck reviewed (dismissed):
The phrase
`an implementation-approximated value representing the result of the <foo> of _n_`
is rather long. I don't think "the result of" is necessary, so I suggest dropping it:
`an implementation-approximated value representing the <foo> of _n_`

Interestingly, there's nothing about "implementation-approximated value" that says it's a Number. Nor is there an over-arching requirement that allows us to infer that the value must be a Number. (The first para of this clause used to say "The value returned by each function is a Number", but this PR removed it.) So maybe reinstate that sentence, or insert "Number":
`an implementation-approximated Number value representing the <foo> of _n_`

Moreover, I think the "representing" is incorrect. The resulting Number value doesn't *represent* `the <foo> of _n_`; rather it represents a mathematical value that approximates `the <foo> of _n_`. So you could maybe say:
`a Number value that implementation-approximates the <foo> of _n_`
but then you don't get auto-linking to "implementation-approximated".

So you could coin a phrase, e.g.
`a Number value that approximates _x_`
or
`a Number-approximation for _x_`
and then define it in 6.1.6.1 "The Number Type", right after defining [`the Number value for _x_`](https://tc39.es/ecma262/#number-value), and mention there that it's implementation-approximated.

Note that an implementation-approximated facility is supposed to recommend "an ideal behaviour". So you could say that for `a Number value that approximates _x_`, the ideal behaviour is `the Number value for _x_`.

bakkot commented:
@jmdyck I agree that the "an implementation-approximated value representing the result of <foo> of _n_" phrasing is awkward and a little imprecise, but its use for math predates this PR; I think we shouldn't worry about changing it here, and instead explore alternative phrasings in a different issue.

jmdyck commented:
> its use for math predates this PR

True, although in preambles rather than algorithm steps.

> I think we shouldn't worry about changing it here, and instead explore alternative phrasings in a different issue.

Okay.

ryanjduffy commented:
Thanks for the feedback, @jmdyck. Fixed up the suggestions other than the discussion on "implementation-approximated value."

bakkot reviewed (approved):
LGTM. Thanks for sticking with it!

ljharb commented on spec.html:
```diff
@@ -28087,278 +28085,158 @@ <h1>Function Properties of the Math Object</h1>
+          1. Return the smallest (closest to *-&infin;*) integral Number value that is not less than _n_.
+        </emu-alg>
+        <emu-note>
+          <p>The value of `Math.ceil(x)` is the same as the value of `-Math.floor(-x)`.</p>
```
i wonder if it'd be simpler to just have this algorithm return `- %Math.floor(-x)%` rather than repeating the steps and having the note.

ljharb commented on spec.html:
```diff
@@ -28502,24 +28329,18 @@ <h1>Math.fround ( _x_ )</h1>
+            1. If _n_ is *NaN*, _n_ is *+&infin;*, or _n_ is *-&infin;*, return _n_.
+            1. If _n_ is neither *+0* nor *-0*, set _onlyZero_ to *false*.
+          1. If _onlyZero_ is *true*, return *+0*.
+          1. Return an implementation-approximated value representing the square root of the sum of squares of the elements of _numbers_.
```
Given that the elements of `numbers` aren't actually all numbers - they might be objects that are `ToNumber`ed - it seems like this actually mandates *two* observable ToNumber calls for any object value.

Obviously this is not the intent, and it's likely that all implementations only call ToNumber once per item, but I think we should lock this down in a future PR.

bakkot replied:
Nice catch. I think we should fix that here, probably by constructing a second list while iterating over the first.

ljharb reviewed (approved):
Fully re-reviewed this; my comments are nonblockers and should be addressed in future PRs.

ljharb commented:
I'm going to rebase and land this now; for any additional feedback, please submit an issue or a PR :-)

ryanjduffy commented:
Thanks all! Good to see it land. 141 comments is definitely a new personal record for a PR :)

michaelficarra commented:
@ryanjduffy Yeah we're pretty thorough here. #2045 is currently sitting at 281 comments. If you want to take up any more open issues, I'd be happy to help out.

ryanjduffy commented:
@michaelficarra - not sure if 281 comments is supposed to be encouraging or intimidating ;)
